stestr/subunit_runner/program.py

```diff
@@ -58,23 +58,31 @@ def filter_by_ids(suite_or_case, test_ids):
     thus the backwards-compatible code paths attempt to mutate in place rather
     than guessing how to reconstruct a new suite.
     """
-    # Compatible objects
-    if extras.safe_hasattr(suite_or_case, 'filter_by_ids'):
-        return suite_or_case.filter_by_ids(test_ids)
-    # TestCase objects.
-    if extras.safe_hasattr(suite_or_case, 'id'):
-        if suite_or_case.id() in test_ids:
-            return suite_or_case
-        else:
-            return unittest.TestSuite()
-    # Standard TestSuites or derived classes [assumed to be mutable].
-    if isinstance(suite_or_case, unittest.TestSuite):
-        filtered = []
-        for item in suite_or_case:
-            filtered.append(filter_by_ids(item, test_ids))
-        suite_or_case._tests[:] = filtered
-    # Everything else:
-    return suite_or_case
+    def _filter_tests(to_filter, id_map):
+        # Compatible objects
+        if extras.safe_hasattr(to_filter, 'filter_by_ids'):
+            res = to_filter.filter_by_ids(test_ids)
+            _filter_tests(res, id_map)
+        # TestCase objects.
+        elif extras.safe_hasattr(to_filter, 'id'):
+            test_id = to_filter.id()
+            if test_id in test_ids:
+                id_map[test_id] = to_filter
+        # Standard TestSuites or derived classes.
+        elif isinstance(to_filter, unittest.TestSuite):
+            for item in to_filter:
+                _filter_tests(item, id_map)
```

Comment:
I'm a little uneasy about the recursion here. I feel like it probably would have been a bit cleaner to just do it iteratively. But it's not a big deal this should be fine. If it becomes an issue we can always refactor it later